### ISSUE TITLE
The heat budget (demand and supply) of the network should be calculated for each time-step

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmoses#971]
This ticket is about the workings of the load calculation for heat. 

* demand: defined by profiles, associated with buffers at the end-points of the electricity topology (nothing new)
* supply: (in order of precedence)
  * **heat network must-run**. These sources supply the heat network with heat. The amount is defined by a yearly budget and the instantaneous production is described by a profile. The production profile for must-run sources is simply the profile scaled with the yearly production (similar to electricity demand profiles). The profile is taken from the load-profiles of the merit order calculation in the ETM, interpolated to 15 minutes. NOTE: The user can choose another profile.
  * **heat network dispatchable sources**. These are also feeding the heat network but only if needed (if demand surpasses must-run supply)
  * **de-central heat sources** (heat pumps, gas-fired heaters etc., these are already in ETMoses) which might be used to fill in the gaps if the heat network (both must-run and dispatchables) have insufficient production capacity to fulfil demand.

Note that because the heat network has a certain buffering volume, say 10 kWh per connection, the must-run sources can input some heat at times when there might not be a matching demand.

So, the must-run sources just fill up the heat network buffer and any remaining heat that cannot be put in the buffer, even after demand is subtracted from the buffer in this time-step, is thrown away (curtailed).

**Comment by ChaelKruip** [comment]
It would be nice if we could show demand and different types of supply per end-node for each time-step.

**Comment by ChaelKruip** [comment]
If we have time, losses can be included (see https://github.com/quintel/etmoses/issues/972)

**Comment by antw** [comment]
> say 10 kWh per connection, the must-run sources can input some heat at times when there might not be a matching demand.

@ChaelKruip We briefly discussed the notion of having two separate buffer volumes. This would be used to model water being heated *beyond* the level needed by endpoints (with that extra-hot water being mixed with cold).

However I'm not sure what the criteria for this extra heat is. Is it that some must-runs can produce the energy needed to reach the higher volume, and some cannot?

For example, if we have a buffer with a "normal" max. volume of 10kW and an "amplified" max. volume *(for want of a better name)* of 15kW, are we saying that some must-runs cannot fill up the buffer if it would exceed 10kW, but other must-runs *can*, up to a maximum of 15kW?

**Comment by ChaelKruip** [comment]
> Is it that some must-runs can produce the energy needed to reach the higher volume, and some cannot?

All must-runs are treated equally and the amount of heat they can dump into the buffer is limited by the `max_volume`. The `normal_volume` is the limiting factor for *dispatchable* sources which are buffering heat into the buffer. 

The rationale behind this is that dispatchable sources usually have worse COP/efficiency if they produce heat at a higher temperature so they are tweaked to fill the buffer with heat up to a temperature that is needed for use by the consumer. For the must-run sources, the COP/efficiency is irrelevant in the sense that their heat is a by-product. So, if their excess heat could not be stored in a buffer it has to be curtailed (thrown out of the window, leading to global warming). In that case it makes sense to ignore any efficiency considerations and fill the buffer the physical max (let's say a bit below the boiling point = 95 degrees Celsius).

Does that make sense?

**Comment by antw** [comment]
> The normal_volume is the limiting factor for dispatchable sources which are buffering heat into the buffer.

Ah, I see. The original post made it seem that dispatchables are purely instantaneous, and only run if demand exceeds the energy stored in the buffer + must-runs. So, dispatchables should also be used to fill up the buffer?

**Comment by ChaelKruip** [comment]
> So, dispatchables should also be used to fill up the buffer?

Yes. Sorry to be unclear about that. I expect the dispatchable sources to behave the same as the buffering technologies in households (without any strategies selected). They fill up the buffer up to the `normal_volume ` after the must-runs have done their thing.

The 'merit order' for heating the heat network is as follows:
* must-run
* dispatchable with highest priority (lower numbers are higher priority)
* dispatchable with second highest priority
* etc.

**Comment by ChaelKruip** [comment]
To be more explicit about the size of the storage volumes: they are based on physical considerations
```
storage_volume = volume_in_liters * heat_capacity_of_water * temperature_difference
```
where the `heat_capacity_of_water = 0.001163 KWh/(liter * K)`.

The `volume_in_liters` can be linked to the volume of the pipes of the heat network which can be assigned to a single connection. The temperature difference would be:
* `temperature_difference_normal = 60 K - 15 K = 45 K`
* `temperature_difference_max = 95 K - 15 K = 80 K`

To end up with 10kWh of buffer volume (normal/dispatchable case), the number of liters of water per connection would need to be:
```
10 kWh / (0.001163 kWh/liter/K * 45K) = 191.1 liter
```
which seems to be within the realm of possibilities (~200 liter per connection).

For the must-run/max case, the equivalent buffer volume would become:
```
191.1 liter * 0.001163 kWh/liter/K * 80 K = 17.78 kWh.
```

@Sanne512 can you check the above reasoning? More specifically:
* Is the volume of 191 liters per connection realistic?
* Is the threshold value of 60 degrees Celsius realistic for the dispatchables to stop heating the network?

**Comment by antw** [comment]
The heat network is now set up such that its (buffering) volume will be set to 10kWh per heat connection. The amplified volume – available only to must-runs – is set to 17.78kWh per connection.
